src/passes/OptimizeInstructions.cpp

```diff
@@ -205,6 +215,182 @@ struct OptimizeInstructions
+      // match API. Once most of the transformations have been ported, the
+      // `using namespace Match` can be hoisted to function scope and this extra
+      // block scope can be removed.
+      using namespace Match;
+      Builder builder(*getModule());
+      {
+        // X == 0 => eqz X
+        Expression* x;
+        if (matches(curr, binary(EqInt32, any(&x), i32(0)))) {
+          return Builder(*getModule()).makeUnary(EqZInt32, x);
+        }
+      }
+      {
+        // try to get rid of (0 - ..), that is, a zero only used to negate an
+        // int. an add of a subtract can be flipped in order to remove it:
+        //   (i32.add
+        //     (i32.sub
+        //       (i32.const 0)
+        //       X
+        //     )
+        //     Y
+        //   )
+        // =>
+        //   (i32.sub
+        //     Y
+        //     X
+        //   )
+        // Note that this reorders X and Y, so we need to be careful about that.
+        Expression *x, *y;
+        Binary* sub;
+        if (matches(curr,
+                    binary(AddInt32,
+                           binary(&sub, SubInt32, i32(0), any(&x)),
+                           any(&y))) &&
+            canReorder(x, y)) {
+          sub->left = y;
+          sub->right = x;
+          return sub;
+        }
+      }
+      {
+        // The flip case is even easier, as no reordering occurs:
+        //   (i32.add
+        //     Y
+        //     (i32.sub
+        //       (i32.const 0)
+        //       X
+        //     )
+        //   )
+        // =>
+        //   (i32.sub
+        //     Y
+        //     X
+        //   )
+        Expression* y;
+        Binary* sub;
+        if (matches(curr,
+                    binary(AddInt32,
+                           any(&y),
+                           binary(&sub, SubInt32, i32(0), any())))) {
+          sub->left = y;
+          return sub;
+        }
+      }
+      {
+        // try de-morgan's AND law,
+        //  (eqz X) and (eqz Y) === eqz (X or Y)
+        // Note that the OR and XOR laws do not work here, as these
+        // are not booleans (we could check if they are, but a boolean
+        // would already optimize with the eqz anyhow, unless propagating).
+        // But for AND, the left is true iff X and Y are each all zero bits,
+        // and the right is true if the union of their bits is zero; same.
+        Unary* un;
+        Binary* bin;
+        Expression *x, *y;
+        if (matches(curr,
+                    binary(&bin,
+                           AndInt32,
+                           unary(&un, EqZInt32, any(&x)),
+                           unary(EqZInt32, any(&y))))) {
+          bin->op = OrInt32;
+          bin->left = x;
+          bin->right = y;
+          un->value = bin;
+          return un;
+        }
+      }
+      // Selects
+      if (auto* select = curr->dynCast<Select>()) {
+        select->condition = optimizeBoolean(select->condition);
+      }
```

Comment:
Actually, I think it would be best to extract these select optimizations into their own function. Like your suggestion, this will also have the effect of having the top-level match compare with a `Select*`, which might save a dynCast after inlining. I definitely want to continue matching from the root of the expression, though, to make it easy to see what is being matched with no context.

Reply:
Wouldn't the name of the variable being "select" be enough to see what is being matched, even if the actual match was a little higher up?

Reply:
(I don't care much, but this does seem like it would be shorter and more efficient the way I'm suggesting, unless I'm missing something.)